// awesome-typescript-loader/dist.babel/test/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.defaultFixturesDir = exports.defaultOutputDir = exports.expect = undefined;
exports.createConfig = createConfig;
exports.expectSource = expectSource;
exports.fixturePath = fixturePath;
exports.readFixture = readFixture;
exports.writeFixture = writeFixture;
exports.touchFile = touchFile;
exports.outputFileName = outputFileName;
exports.readOutputFile = readOutputFile;
exports.cleanOutputDir = cleanOutputDir;
exports.cleanAndCompile = cleanAndCompile;
exports.compile = compile;
exports.watch = watch;

var _path = require('path');

var path = _interopRequireWildcard(_path);

var _fs = require('fs');

var fs = _interopRequireWildcard(_fs);

var _lodash = require('lodash');

var _ = _interopRequireWildcard(_lodash);

var _chai = require('chai');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

require('babel-polyfill');
require('source-map-support').install();
exports.expect = _chai.expect;

var webpack = require('webpack');
var BPromise = require('bluebird');
var mkdirp = BPromise.promisify(require('mkdirp'));
var rimraf = BPromise.promisify(require('rimraf'));
var readFile = BPromise.promisify(fs.readFile);
var writeFile = BPromise.promisify(fs.writeFile);
var loaderDir = path.join(process.cwd(), 'dist.babel');
var ForkCheckerPlugin = require(loaderDir).ForkCheckerPlugin;
var defaultOutputDir = exports.defaultOutputDir = path.join(process.cwd(), 'src', 'test', 'output');
var defaultFixturesDir = exports.defaultFixturesDir = path.join(process.cwd(), 'src', 'test', 'fixtures');
var defaultOptions = {
    watch: false,
    forkChecker: false
};
function createConfig(conf) {
    var _options = arguments.length <= 1 || arguments[1] === undefined ? defaultOptions : arguments[1];

    var options = _.merge({}, defaultOptions, _options);
    var defaultConfig = {
        watch: false,
        output: {
            path: defaultOutputDir,
            filename: '[name].js'
        },
        resolve: {
            extensions: ['', '.ts', '.tsx', '.js', '.jsx']
        },
        module: {
            loaders: [{
                test: /\.(tsx?|jsx?)/,
                loader: loaderDir,
                query: Object.assign({ target: 'es6' }, options.loaderQuery)
            }]
        },
        plugins: []
    };
    if (options.include) {
        defaultConfig.module.loaders[0].include = options.include;
    }
    if (options.exclude) {
        defaultConfig.module.loaders[0].exclude = options.exclude;
    }
    if (options.watch) {
        defaultConfig.watch = true;
    }
    if (options.forkChecker) {
        defaultConfig.plugins.push(new ForkCheckerPlugin());
    }
    return _.merge(defaultConfig, conf);
}
function expectSource(source, fragment) {
    (0, _chai.expect)(source.replace(/\s/g, '')).include(fragment.replace(/\s/g, ''));
}
function fixturePath(fileName) {
    var fixturesDir = arguments.length <= 1 || arguments[1] === undefined ? defaultFixturesDir : arguments[1];

    return path.join.apply(path, [fixturesDir].concat(fileName));
}
function readFixture(fileName) {
    var fixturesDir = arguments.length <= 1 || arguments[1] === undefined ? defaultFixturesDir : arguments[1];

    var filePath = fixturePath(fileName, fixturesDir);
    return readFile(filePath).then(function (buf) {
        return buf.toString();
    });
}
function writeFixture(fileName, text) {
    var fixturesDir = arguments.length <= 2 || arguments[2] === undefined ? defaultFixturesDir : arguments[2];

    var filePath = fixturePath(fileName, fixturesDir);
    return writeFile(filePath, text);
}
function touchFile(fileName) {
    return readFile(fileName).then(function (buf) {
        return buf.toString();
    }).then(function (source) {
        return writeFile(fileName, source);
    });
}
function outputFileName(fileName) {
    var outputDir = arguments.length <= 1 || arguments[1] === undefined ? defaultOutputDir : arguments[1];

    return path.join(defaultOutputDir, fileName);
}
function readOutputFile(fileName) {
    var outputDir = arguments.length <= 1 || arguments[1] === undefined ? defaultOutputDir : arguments[1];

    return readFile(outputFileName(fileName || 'main.js', outputDir)).then(function (buf) {
        return buf.toString();
    });
}
function cleanOutputDir() {
    var outputDir = arguments.length <= 0 || arguments[0] === undefined ? defaultOutputDir : arguments[0];

    return rimraf(outputDir).then(function () {
        return mkdirp(outputDir);
    });
}
function cleanAndCompile(config) {
    var outputDir = arguments.length <= 1 || arguments[1] === undefined ? defaultOutputDir : arguments[1];

    return cleanOutputDir(outputDir).then(function () {
        return compile(config);
    });
}
function compile(config) {
    return new Promise(function (resolve, reject) {
        var compiler = webpack(config);
        compiler.run(function (err, stats) {
            if (err) {
                reject(err);
            } else {
                resolve(stats);
            }
        });
    });
}
function watch(config, cb) {
    var compiler = webpack(config);
    return new Promise(function (resolve, reject) {
        var watcher = compiler.watch({}, function (err, stats) {
            if (err) {
                reject(err);
            } else {
                resolve(watcher);
            }
            cb(err, stats);
        });
    });
}
